pkg/acsengine/orchestrators.go

```diff
@@ -0,0 +1,273 @@
+	if len(arr) != 3 {
+		return fmt.Errorf("Illegal version format '%s'", ver)
+	}
+	if v.major, err = strconv.ParseInt(arr[0], 10, 32); err != nil {
+		return
+	}
+	if v.minor, err = strconv.ParseInt(arr[1], 10, 32); err != nil {
+		return
+	}
+	if v.patch, err = strconv.ParseInt(arr[2], 10, 32); err != nil {
+		return
+	}
+	return nil
+}
+
+func (v *versionNumber) greaterThan(o *versionNumber) bool {
+	// check major
+	if v.major != o.major {
+		return v.major > o.major
+	}
+	// same major; check minor
+	if v.minor != o.minor {
+		return v.minor > o.minor
+	}
+	// same minor; check patch
+	return v.patch > o.patch
+}
+
+func init() {
+	funcmap = map[string]orchestratorsFunc{
+		api.Kubernetes: kubernetesInfo,
+		api.DCOS:       dcosInfo,
+		api.Swarm:      swarmInfo,
+		api.SwarmMode:  dockerceInfo,
+	}
+}
+
+func validate(orchestrator, release string) (string, error) {
+	switch {
+	case strings.EqualFold(orchestrator, api.Kubernetes):
+		return api.Kubernetes, nil
+	case strings.EqualFold(orchestrator, api.DCOS):
+		return api.DCOS, nil
+	case strings.EqualFold(orchestrator, api.Swarm):
+		return api.Swarm, nil
+	case strings.EqualFold(orchestrator, api.SwarmMode):
+		return api.SwarmMode, nil
+	case len(orchestrator) == 0:
+		if len(release) > 0 {
+			return "", fmt.Errorf("Must specify orchestrator for release '%s'", release)
+		}
+	default:
+		return "", fmt.Errorf("Unsupported orchestrator '%s'", orchestrator)
+	}
+	return "", nil
+}
+
+// NewOrchestratorInfos returns OrchestratorInfos object per (optionally) specified orchestrator and release
+func NewOrchestratorInfos(orchestrator, release string) (*OrchestratorInfos, error) {
+	var err error
+	if orchestrator, err = validate(orchestrator, release); err != nil {
+		return nil, err
+	}
+	orch := &OrchestratorInfos{}
+
+	if len(orchestrator) == 0 {
+		// return all orchestrators
+		orch.Orchestrators = []*api.OrchestratorInfo{}
+		for _, f := range funcmap {
+			arr, err := f(&api.VersionInfo{})
+			if err != nil {
+				return nil, err
+			}
+			orch.Orchestrators = append(orch.Orchestrators, arr...)
+		}
+		return orch, nil
+	}
+	if orch.Orchestrators, err = funcmap[orchestrator](&api.VersionInfo{Release: release}); err != nil {
+		return nil, err
+	}
+	return orch, nil
+}
+
+// GetOrchestratorUpgradeInfo returns orchestrator info for upgradable container service
+func GetOrchestratorUpgradeInfo(cs *api.ContainerService) (*api.OrchestratorInfo, error) {
+	if cs == nil || cs.Properties == nil || cs.Properties.OrchestratorProfile == nil {
+		return nil, fmt.Errorf("Incomplete ContainerService")
+	}
+	if len(cs.Properties.OrchestratorProfile.OrchestratorRelease) == 0 {
+		return nil, fmt.Errorf("Missing Orchestrator Release")
+	}
+	if cs.Properties.OrchestratorProfile.OrchestratorType != api.Kubernetes {
+		return nil, fmt.Errorf("Upgrade operation is not supported for '%s'", cs.Properties.OrchestratorProfile.OrchestratorType)
+	}
+	arr, err := kubernetesInfo(&api.VersionInfo{
```

Comment:
As noted in general it's not clear why another struct is needed to denote the same info as orchestratorProfile. This looks like a duplication that can be avoided.